### PR TITLE
Added request interceptor pattern

### PR DESCRIPTION
**Pull request by fehguy**
This addresses #599 as well as #591.  With this PR, you can now intercept *all* requests and responses with the same pattern, allowing auditing request flows by creating timers, adding arbitrary headers, etc.  Note unlike the signing mechanisms, this does _not_ require a `security` tag to match a custom signing algorithm.  You should still use the security tags for selective signing of calls, and the interceptors for handling *all* calls.

@marcellodesales please review to ensure this addresses your needs, this implementation is more consistent with the swagger-js project.

### CODE REVIEW

**Comment by marcellodesales** [comment]
@fehguy I will take a look at it and make sure it works in our environment! Thanks a lot for considering it!

**Comment by marcellodesales** [comment]
@fehguy Can you consolidate this branch?

**Comment by fehguy** [comment]
sure @marcellodesales just rebased

**Comment by marcellodesales** [comment]
@fehguy perfect! thanks!

File: lib/http.js [hunk 1]
```diff
@@ -48,6 +49,13 @@ SwaggerHttp.prototype.execute = function (obj, opts) {
...
+  var requestInterceptor = function(data) {
+    if(opts && opts.requestInterceptor) {
+      data = opts.requestInterceptor.apply(data);
+    }
+    return data;
+  }
+
   var responseInterceptor = function(data) {
     if(opts && opts.responseInterceptor) {
       data = opts.responseInterceptor.apply(data);
```
Comment (marcellodesales) [inline lib/http.js]: How would I get access to the `req`, `res`, and `err` objects to support #599 ???

File: test/request.js [hunk 2]
```diff
@@ -405,6 +405,69 @@ describe('swagger basePath override functions', function () {
...
+      done();
+    }, function(err){
+      expect(err.obj).toBeAn('object');
+      expect(err.obj.code).toBe(400);
+      expect(err.obj.type).toBe('bad input');
+      expect(err.obj.message).toBe('sorry!');
+      done();
+    });
+  });
+
```
Comment (marcellodesales) [inline test/request.js]: Can you please add a supporting test case for #599 as you claim this patch solves for it??? I still cannot see how the API has access to the lower-level `req`, `res`, and `err` objects.

**Comment by marcellodesales** [comment]
@fehguy I cannot understand how this would solve #599 ... Could you please address the questions above and add a test case to support your PR? Thanks!

**Pull request merged**